### PR TITLE
Brownian motion updated algorithm, with merge conflict

### PR DESCRIPTION
**Pull request by andrlime**

### CODE REVIEW

File: exchange/src/randomness/brownian.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ */
+
+#pragma once
+
+#define BROWNIAN_MOTION_MEAN_SIZE_EVENT  15
+#define BROWNIAN_MOTION_STDEV_EVENT_SIZE 5
+#define BROWNIAN_MOTION_DEVIATION        0.1
+#define SKEW_SCALE                       20000
+#define SKEW_FACTOR                      4
```
Comment (egelja) [inline exchange/src/randomness/brownian.hpp:18]: this should probably go into `config.h`

File: exchange/src/randomness/brownian.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ */
+
+#include "brownian.hpp"
+
+namespace nutc {
+namespace stochastic {
+
+double
+BrownianMotion::generate_next_price()
```
Comment (egelja) [inline exchange/src/randomness/brownian.cpp:18]: i think this probably count be refactored into a couple of helper functions

File: exchange/src/randomness/brownian.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    // Not ticking, so we need a random number
+    std::uniform_real_distribution<double> distribution(0.0, 1.0);
+    double random_number = distribution(random_number_generator_);
+
+    if (random_number >= probability_) {
+        std::normal_distribution<double> distribution(
+            BROWNIAN_MOTION_MEAN_SIZE_EVENT, BROWNIAN_MOTION_STDEV_EVENT_SIZE
+        );
+        ticker_ = (int)abs(distribution(random_number_generator_));
```
Comment (egelja) [inline exchange/src/randomness/brownian.cpp:43]:
```suggestion
        ticker_ = static_cast<size_t>( //
            abs(distribution(random_number_generator_))
        );
```

File: exchange/src/randomness/brownian.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        // Whether to tick up or tick down
+        ticking_up_ = distribution(random_number_generator_) > 0.5;
+
+        // Generate change in price
+        std::normal_distribution<double> distribution(
+            -cur_value_ / skew_scale_, BROWNIAN_MOTION_DEVIATION
+        );
+        double delta_current_price = distribution(random_number_generator_);
+        double new_price = new_price + delta_current_price;
+        cur_value_ = new_price;
```
Comment (egelja) [inline exchange/src/randomness/brownian.cpp:54]: this code is duped 3x you could probably extract it into a helper

File: exchange/src/randomness/brownian.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace nutc {
+namespace stochastic {
+
+double
+BrownianMotion::generate_next_price()
+{
+    if (ticker_ > 0) {
+        ticker_--;
+        int sign = ticking_up_ ? 1 : -1;
```
Comment (egelja) [inline exchange/src/randomness/brownian.cpp:22]: this variable seems to be unused

**Review by egelja — CHANGES_REQUESTED** [review]
looks good, just some nitpicks

File: exchange/src/randomness/brownian.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+double
+BrownianMotion::generate_change_in_price(double mean, double stdev, Signedness sign)
+{
+    std::normal_distribution<double> distribution(mean, stdev);
+
+    if (sign == Signedness::DoesntMatter) {
+        return distribution(random_number_generator_);
+    }
+    else {
+        short multiplier = (sign == Signedness::Positive ? 1 : -1);
```
Comment (egelja) [inline exchange/src/randomness/brownian.cpp:26]:
```suggestion
        int multiplier = static_cast<int>(sign);
```

File: exchange/src/randomness/brownian.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        return multiplier * abs(distribution(random_number_generator_));
+    }
+}
+
+double
+BrownianMotion::generate_next_price()
+{
+    if (ticker_ > 0) {
+        ticker_--;
+        Signedness sign = ticking_up_ ? Signedness::Positive : Signedness::Negative;
```
Comment (egelja) [inline exchange/src/randomness/brownian.cpp:36]: why not just store the signedness as a class field?

File: exchange/src/randomness/brownian.hpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    // Constructor for BrownianMotion, takes a seed
+    explicit BrownianMotion(const unsigned int seed) : cur_value_(0)
+    {
+        random_number_generator_ = std::mt19937(seed);
+    }
+
+    // Generates and returns the change in price, i.e. dp/dt
+    [[nodiscard]] double
+    generate_change_in_price(double mean, double stdev, Signedness sign);
```
Comment (egelja) [inline exchange/src/randomness/brownian.hpp:48]: should probably be private

File: exchange/src/randomness/brownian.hpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    explicit BrownianMotion() : cur_value_(0)
+    {
+        std::random_device rd;
+        random_number_generator_ = std::mt19937(rd());
+    }
+
+    // Constructor for BrownianMotion, takes a seed
+    explicit BrownianMotion(const unsigned int seed) : cur_value_(0)
+    {
+        random_number_generator_ = std::mt19937(seed);
```
Comment (egelja) [inline exchange/src/randomness/brownian.hpp:43]:
```suggestion
        set_seed(seed);
```

File: exchange/src/randomness/brownian.cpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        // Generate new price
+        cur_value_ += generate_change_in_price(
+            -cur_value_ / SKEW_SCALE, BROWNIAN_MOTION_DEVIATION,
+            Signedness::DoesntMatter
+        );
+        return cur_value_;
+    }
+    else {
+        cur_value_ += generate_change_in_price(
+            -cur_value_ / SKEW_SCALE, BROWNIAN_MOTION_DEVIATION / SKEW_FACTOR,
```
Comment (egelja) [inline exchange/src/randomness/brownian.cpp:65]: why is this the only place you use `SKEW_FACTOR`?

**Review by egelja — CHANGES_REQUESTED** [review]

File: exchange/src/randomness/brownian.cpp (hunk 10 above)
Comment (andrlime) [reply exchange/src/randomness/brownian.cpp:65]: I use this to amplify the magnitude of the market shifting events as opposed to non-market shifting events to make them more dramatic

**Review by echavemann — COMMENTED** [review]
brownian looks fine to me

File: exchange/src/randomness/brownian.cpp (hunk 10 above)
Comment (egelja) [reply exchange/src/randomness/brownian.cpp:65]: oh ok just double checking.

File: exchange/src/logging.hpp [hunk 11]
```diff
@@ -117,4 +119,4 @@ CREATE_LOG_CATEGORY(sandbox);
 
 #define log_c(category, ...)                                                           \
     LOG_CRITICAL(nutc::logging::get_##category##_logger(), __VA_ARGS__)
-// NOLINTEND
+// NOLINTEND
```
Comment (egelja) [inline exchange/src/logging.hpp:122]: pls include newline thx

File: exchange/src/randomness/brownian.hpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    void
+    set_seed(unsigned int new_seed)
+    {
+        random_number_generator_ = std::mt19937(new_seed);
+    }
+
+private:
+    // Generates and returns the change in price, i.e. dp/dt
+    [[nodiscard]] double
+    generate_change_in_price(double mean, double stdev, Signedness sign);
```
Comment (egelja) [inline exchange/src/randomness/brownian.hpp:73]:
```suggestion
    generate_change_in_price_(double mean, double stdev, Signedness sign);
```

**Review by egelja — APPROVED** [review]
lgtm

**Review by egelja — APPROVED** [review]
lgtm

**Pull request closed without merging**